decidim-core/app/assets/javascripts/decidim/vizzs/metrics.js.es6

```diff
@@ -1,26 +1,20 @@
 /* eslint-disable no-unused-vars */
 
-// Outside of the closure to make it public
-let DATACHARTS = null;
-
-const fetchDatacharts = () => {
-
-  const metrics = [{
-    key: "NAME_TO_BE_IN_THE_HTML-1",
-    query: "GRAPHQL_QUERY-1"
-  }, {
-    key: "NAME_TO_BE_IN_THE_HTML-2",
-    query: "GRAPHQL_QUERY-2"
-  }]
-
-  const fetch = (query) => $.post("<-- GRAPHQL_URL -->", query)
-
-  const promises = metrics.map((metric) => fetch(metric.query).then((response) => {
-    DATACHARTS[metric.key] = response
-
-    return DATACHARTS
-  }))
-
-  Promise.all(promises).then(() => DATACHARTS)
-
-}
+$(() => {
+  const fetch = (metric) => $.post("api", query(metric))
+
+  const query = (metric) => {
+    return {query: `{ ${metric} { metric { key value } } }`};
+  }
+
+  $(".metric-chart:visible").each((i, container) => {
+    let metric = $(container).data("metric");
+    fetch(metric).then((response) => {
+      areachart({
```

Comment:
Take into account we must render different chart types (currently we manage linechart, rowchart, areachart, and orgchart). So you must distinguish each type in order to render one or another.

Reply:
I'm aware of this. This is still a WIP PR; more charts I'll show, more complex will be this JS

Reply:
Look at [this](https://github.com/decidim/decidim/pull/3853/files#diff-4bed86a86ba52335a884096cc2fb60e8).
I think it's a good approach to separate the `render` function (it doesn't know what is rendering) from the chart itself. You may call it, running: 
```js
$(() => {
    render(renderOrgCharts);
})
```
where `renderOrgCharts` is the private method who manages the logic & display; and `render` is the function which runs the callback you're passing in the args.

In that way, each chart will know what to render, getting the API url (w/ its params) from the metric attribute.

Reply:
Each `render` must wait to the API response. Can we do it with this?

Reply:
Yep, with `Promises`.
But currently, the app is exposing to the `window` object all the stuff it needs to share (You'll notice it when in some scripts you see `exports.Decidim`, or similar).

That's "OK" (I remark, quotes) for sharing objects or functions, but to expose the endpoint url plus object-like Graphql params... mmm I'm not quite sure about that (too risky, it's sensitive data) cc/ @decidim/lot-core

But we can declare it in the global js scope, as the `DATACHARTS` object, so once promise is solved (`fetchDataCharts()`), we can call the charts render function

Reply:
No, it's not sensitive data. And currently Graphql is public open, so no problem to expose the endpoint or params

Reply:
In such case, we can store different objects in `Decidim.Visualizations`, for instance:
```
Decidim.Visualizations.Areachart.metric-1
Decidim.Visualizations.Orgchart.metric-45
Decidim.Visualizations.Linechart.metric-12
...
```
`Decidim.Visualizations.CHART-TYPE.METRIC-NAME`, where metric-name is what the HTML data-attribute has, and each content are the graphql params.

Reply:
I cannot see how to handle this. Can you show me an example, please?

Reply:
Look [at the previous way](https://github.com/decidim/decidim/pull/3603/files/fbdcac360dbd11c57a0ac1fc375d36c0c752796b#diff-5bd70f01d854221087856af79a226aeaL16) to populate `DATACHARTS` variable